Add padEnds option to SliverFillViewport

mayabyte opened the pull request:
## Description

Adds the `padEnds` option to SliverFillViewport. 

`padEnds` configures whether or not padding should be added to the end of the SliverFillViewport to make the first and last elements be centered when scrolled all the way to one end of the list. Previously this padding was always added, but in some cases it can be undesirable. Setting `padEnds` to `false` removes this padding entirely. `padEnds` defaults to `true` to ensure this change doesn't break any existing code.

### Example Use

I encountered this padding when writing a calendar application. I'm using the trick of having two opposite-direction lists (in this case SliverFillViewports with `viewportFraction=1/5`, because I want the viewport to show 5 weeks) in a viewport to create an infinitely-scrolling list of weeks in both directions. However, the padding added by SliverFillViewport creates an unseemly gap between the two lists:

<img width="400" alt="image" src="https://user-images.githubusercontent.com/42569602/71780458-80d3ad00-2fba-11ea-8158-a512f64308bb.png">

This is the desired appearance, achieved by setting `padEnds` to `false`:

<img width="400" alt="image" src="https://user-images.githubusercontent.com/42569602/71780504-3e5ea000-2fbb-11ea-9612-1402e650f304.png">

## Related Issues

https://github.com/flutter/flutter/issues/48252

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

goderbauer commented:
/cc @Piinks

Piinks commented:
Hi @mayabyte, welcome! And thank you for the contribution!  🎉 
Would you be willing to open an issue to reference here? Also, If you can provide a minimal code sample with the issue that reproduces the bug we can determine if this is the appropriate fix for it.

Thanks!

Also, cc'ing @LongCatIsLooong It looks like you have worked in this area recently.

mayabyte commented:
Sure! I'll write up a minimal code sample and include it in an issue

mayabyte commented:
https://github.com/flutter/flutter/issues/48252

LongCatIsLooong reviewed (commented):
Thank you for contributing! I think this is definitely a nice addition to `SliverFillViewport`.

[Our policy](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#write-test-find-bug) states that pull requests like this must include tests. Could you add a few simple tests that verify the new parameter works?

mayabyte commented:
Looks like there was some refactoring done in the meantime. I'll fix that up

mayabyte committed:
Add padEnds option to SliverFillViewport

mayabyte commented:
Took me a while to figure out how testing in Flutter works but I managed to put together something that seems okay 😄

Also went and updated the doc comments as you suggested and added the assert

Piinks commented on packages/flutter/lib/src/widgets/sliver_fill.dart:
```diff
@@ -41,13 +43,26 @@ class SliverFillViewport extends StatelessWidget {
+  /// children.
+  ///
+  /// This option cannot be [null]. If [viewportFraction] >= 1.0, no padding will be
+  /// added regardless, so this option does nothing. Defaults to [true].
```
The second sentence is a little confusing, does this say what you mean?
```dart
/// This option cannot be [null]. If [viewportFraction] >= 1.0, this option has no effect.
/// Defaults to [true].
```

mayabyte replied:
I like that a bit better 😄

Piinks commented on packages/flutter/lib/src/widgets/sliver_fill.dart:
```diff
@@ -41,13 +43,26 @@ class SliverFillViewport extends StatelessWidget {
   Widget build(BuildContext context) {
     return _SliverFractionalPadding(
-      viewportFraction: (1 - viewportFraction).clamp(0, 1) / 2,
+      viewportFraction: padEnds ? (1 - viewportFraction).clamp(0, 1) / 2 : 0,
```
I wonder if it might be more efficient to just not include the `_SliverFractionalPadding` if `!padEnds`. Providing a padding of 0 essentially make it a no-op widget I think. WDYT @LongCatIsLooong?

mayabyte replied:
I opted not to do that initially, mostly because I'm fairly new to Flutter and I'm not 100% familiar how changing the internal structure of widgets impacts other things down the line. If we can be reasonably sure it won't affect anything, I'd definitely prefer to omit the `_SliverFractionalPadding` entirely

LongCatIsLooong replied:
Sounds good! My only concern is if `SliverFillViewport` has a stateful child, toggling `padEnds` may cause the stateful child to lose state. But that doesn't sound like a valid use case.

Piinks replied:
I concur! :)
The potential state issue could be resolved by adding a getter/setter for `padEnds` that then marks the widget to update the state.

mayabyte replied:
Correct me if I'm wrong, but since `SliverFillViewport` is a `StatelessWidget`, isn't it supposed to be invalid to mutate its state after creation? So toggling `padEnds` wouldn't even be allowed, right?

LongCatIsLooong replied:
Oh sorry by "toggle" what I really meant was to rebuild the `SliverFillViewport` with a different `padEnds` value, e.g.:
```dart
class SomeStatefulWidget extends StatefulWidget {
  bool padEnds;

  @override
  Widget build(BuildContext context) => SliverFillViewport(padEnds: padEnds);
}   
```

Piinks replied:
I guess it depends on how it is used! I have seen a similar issue pop up with a stateless widget that was used in the composition of a stateful widget that toggled a property.

mayabyte replied:
Ah, that makes more sense. So given that there's a workaround and it's probably an uncommon use-case, I'll have it omit the `_SliverFractionalPadding`

goderbauer replied:
I think it would have been better to not change the shape of the tree when padEnds is toggled from true to false or vice versa as that may result in some state loss in stateful widgets.

mayabyte replied:
Undid that change as it seems like it caused a post-commit test to fail anyway

Piinks reviewed (commented):
Thanks for the update @mayabyte!

LongCatIsLooong commented on packages/flutter/test/widgets/sliver_fill_viewport_test.dart:
```diff
@@ -145,4 +146,50 @@ void main() {
     );
   });
+
+  testWidgets('SliverFillViewport padding test', (WidgetTester tester) async {
```
:+1:

LongCatIsLooong commented on packages/flutter/test/widgets/sliver_fill_viewport_test.dart:
```diff
@@ -145,4 +146,50 @@ void main() {
+      ),
+    );
+
+    final RenderSliver boxWithPadding = tester.renderObject<RenderSliver>(find.byType(SliverFillViewport).first);
```
nit: is there more than one `SliverFillViewport` widget? I think removing the `.first` should be fine? 
```suggestion
    final RenderSliver boxWithPadding = tester.renderObject<RenderSliver>(find.byType(SliverFillViewport));
```

mayabyte replied:
Didn't know you could do that :P

LongCatIsLooong reviewed (approved):
LGTM. Thank you for reporting the issue and providing a fix!

mayabyte committed:
Remove unnecessary '.first' from test

Co-Authored-By: LongCatIsLooong <31859944+LongCatIsLooong@users.noreply.github.com>

goderbauer commented on packages/flutter/lib/src/widgets/sliver_fill.dart:
```diff
@@ -41,18 +43,36 @@ class SliverFillViewport extends StatelessWidget {
+        viewportFraction: (1 - viewportFraction).clamp(0, 1) / 2,
+        sliver: internal,
+      );
+    } else {
```
remove the else and just `return internal;`

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/72096610-67d73e80-32cf-11ea-92a7-498b57360d81.png)

mayabyte commented:
Thanks for the reviews and guidance everyone 😄
